diff --git a/src/core/Program.cpp b/src/core/Program.cpp
index 93b28d9..98639ec 100644
--- a/src/core/Program.cpp
+++ b/src/core/Program.cpp
@@ -430,7 +430,7 @@ bool Program::build(BuildType buildType, const char* options,
         if ((dirend = strrchr(libpath, '/')))
 #endif
         {
-          const char* includes_relative = "/../include/oclgrind/";
+          const char* includes_relative = "/../../oclgrind_binary_distribution/.libs/include/";
           size_t length = dirend - libpath;
           pchdir = new char[length + strlen(includes_relative) + 1];
           strncpy(pchdir, libpath, length);
